Run Invokes during New, not during Start

glibsm opened the pull request:
This builds on top of a lot of previous work, most notably the refactor
of the top level API to no longer expose `Provide` and `Invoke` on the
app object. Now that all options have to atomically be provided on
`New`, it no longer makes sense to wait until the application start to
resolve.

All the heavy-weight work, like listening for the network, will still be
done as part of the `OnStart` hooks.

Fixes #550.

Essentially this introduces a new touch point: when the graph is fully
resolved, but the start Hooks have not yet run. I see some potential
downsides of this, such as metrics collectors not being started and
someone trying to interact with them...

breerly commented on app.go:
```diff
@@ -138,7 +138,13 @@ type App struct {
+// supplied by the Provide option. If any invoked function fails, an error is
+// returned immediately.
 func New(opts ...Option) *App {
+	// TODO(glib): New should probably return (*App, error)
```
`New` should be easy as possible to use - lets just panic instead of returning an error, especially since there is nothing that can be done to recover.

alsamylkin replied:
Why? 
1. It is so much easier to follow the same pattern as the constructors.
2. Panic means that something went really wrong, so we should do it when the internal logic is inconsistent, users invoke's can easily return errors.

Another solution could be provide to an option, to skip/panic/exit on errors(like pflag: https://github.com/spf13/pflag/blob/master/flag.go#L119)

breerly replied:
Mainly because there is nothing that can be recovered, so this just makes the API harder to use.

Let's get @akshayjshaw's opinion.

akshayjshah replied:
Think of a panic as a marginally more polite segfault; just because there's nothing that the user can do to fix this app object doesn't mean that we shouldn't shut the process down gracefully. At the very least, any other long-lived goroutines should have the chance to drain, checkpoint their state, and exit. `Run` gets to cheat, but `New`, `Start`, `Done`, and `Stop` should behave themselves and let the service owner decide how to handle errors.

I'm not opposed to returning an error from `New`, but we don't necessarily need to.

We can instead add an `Err` method that returns `app.optionErr` (which we should rename to `app.err`). This is the same approach used to keep [`bufio.Scanner`](https://godoc.org/bufio#Scanner) easy to use without resorting to panics. It should let 

```
fx.New(...some options...).Run()
```
continue to work, since `Start` can continue to short-circuit if `New` failed.

Users who go the fully-manual route have to handle the error somewhere, so they may end up with

```
app := fx.New(...options...)
if err := app.Err(); err != nil {
  log.Fatalf(err)
}
...usual context creation/Start/Done/Stop dance...
```
If they're just going to call `Start` right away, though, they can skip the explicit `Err` check.

alsamylkin replied:
My last resort :) https://github.com/golang/go/wiki/PanicAndRecover

By convention, no explicit panic() should be allowed to cross a package boundary. Indicating error conditions to callers should be done by returning error value.

breerly commented on app.go:
```diff
@@ -153,6 +159,24 @@ func New(opts ...Option) *App {
+		return nil
+	}
+
+	// Execute invokes.
```
Lets move this entire block into a private func `App.invoke`.

breerly commented on app.go:
```diff
@@ -153,6 +159,24 @@ func New(opts ...Option) *App {
+		fname := fxreflect.FuncName(fn)
+		app.logger.Printf("INVOKE\t\t%s", fname)
+		if err := app.container.Invoke(fn); err != nil {
+			app.logger.Printf("Error during %q invoke: %v", fname, err)
```
Use `app.logger.Fatalf()` here.

akshayjshah replied:
See above.

breerly commented on app.go:
```diff
@@ -153,6 +159,24 @@ func New(opts ...Option) *App {
+
+	if app.optionErr != nil {
+		// Some provides failed, short-circuit immediately.
+		app.logger.Printf("Error after options were applied: %v", app.optionErr)
```
use `app.logger.Fatalf()` here

akshayjshah replied:
See above.

breerly commented on app.go:
```diff
@@ -153,6 +159,24 @@ func New(opts ...Option) *App {
+	}
+
+	// Execute invokes.
+	// TODO(glib): wrap in timeout?
```
Shouldn't need to. If we run into this in the future we can add it though.

akshayjshah commented on app.go:
```diff
@@ -173,16 +197,11 @@ func (app *App) Run() {
 }
 
-// Start starts the application.
+// Start executes all the onStart hooks of the fx.Lifecycle.
```
Can we explain what this means to a typical service owner, who probably won't ever interact with lifecycle? At the least, we can add a second sentence: "This typically starts long-running goroutines, like network servers or message queue consumers."

Small nit: `onStart` should have initial caps to match the `Hook` field name.

glibsm replied:
Yeah. I was trying to cut some corners for the sake of opening this PR, totally agree on making this godoc better.

akshayjshah commented on app.go:
```diff
@@ -224,19 +243,6 @@ func (app *App) provide(constructor interface{}) {
-	if app.optionErr != nil {
-		// Some provides failed, short-circuit immediately.
-		return app.optionErr
-	}
```
Start should still short-circuit if `New` encountered an error.

glibsm commented:
Apologies for long-running PR. Cleaning up right now and merging with all the changes that went in

breerly commented:
@glibsm I've been working under the impression that this is WIP. 

Once it's ready for review, give us a heads up.

breerly commented on app.go:
```diff
@@ -193,9 +198,43 @@ func New(opts ...Option) *App {
 	return app
 }
 
+// runInvokes will execute invokes in order.
```
"in the order provided"

glibsm replied:
That was the original wording, but I didn't want to use the word "provide" to describe part of the "invoke" functionality related to the order.

glibsm commented on inject_test.go:
```diff
@@ -204,33 +203,6 @@ func TestInject(t *testing.T) {
 		assert.True(t, gave3 == out.T3, "T3 must match")
 	})
 
-	t.Run("OverwritesExisting", func(t *testing.T) {
```
Note that this test no longer applies. It was trying to overwrite a value of an already-injected struct. 

This no longer applies since Start no longer modifies the graph, it's fully resolved.

breerly replied:
This is actually a really positive sign for this PR.

breerly commented on app.go:
```diff
@@ -193,9 +198,43 @@ func New(opts ...Option) *App {
 	return app
 }
 
+// Execute invokes in order.
```
Perhaps "in the order supplied" is a little more clear

glibsm replied:
There we go

glibsm commented:
@breerly It was WIP, but still not cool to leave a looming PR for a week.

It's now ready to review, I've addressed all the comments from the discussions and prototypes. The resulting LOC change is actually quite small.

breerly commented on app.go:
```diff
@@ -193,9 +198,43 @@ func New(opts ...Option) *App {
+			return err
+		}
+	}
+
```
nit - space likely unneeded here

breerly commented on app.go:
```diff
@@ -213,16 +252,15 @@ func (app *App) Run() {
 
-// Start starts the application.
+// Start executes all the OnStart hooks of the resolved object graph
+// in the intantiation order.
```
instantiation*

bayesianmind commented on app.go:
```diff
@@ -193,9 +198,43 @@ func New(opts ...Option) *App {
+		app.logger.Printf("INVOKE\t\t%s", fname)
+
+		if _, ok := fn.(Option); ok {
+			return fmt.Errorf("fx.Option should be passed to fx.New directly, not to fx.Invoke: fx.Invoke received %v", fn)
```
nice, appreciate the rough edge detection here!

alsamylkin replied:
why this error is not logged, but the next one is?

bayesianmind commented on app.go:
```diff
@@ -193,9 +198,43 @@ func New(opts ...Option) *App {
+	}
+
+	if err := app.executeInvokes(); err != nil {
+		app.err = err
```
I'd really like to be able to get at this err without having to call Start()/trigger lifecycle.
Ideas:
- fx.TryNew() method which returns (*App, error)
- fx.Option that is an error callback
- app.Error() method

breerly replied:
I really dont want to return an err from New if possible.

I'm OK with the idea of `fx.TryNew` or `app.Error`.

@glibsm @akshayjshah @abhinav can you weigh in here?

abhinav replied:
I like `app.Err` better than `TryNew`.

```
app := New(..)
app.Run()
```

Is equivalent to,

```
app := New(..)

if err := app.Err(); err != nil {
  panic(err)
}

// Usual Start/Stop logic
```

`Start` would need to account for `Err` not having been checked manually by the user.

(Also, let's call it `Err()` instead. There's precedent: https://golang.org/pkg/bufio/#Scanner.Err)

alsamylkin replied:
I am a bit confused which errors are logged and which are not.

glibsm replied:
I will clean up so everything is logged.

breerly replied:
Nice, look like we have `Err()` now :)

breerly replied:
Does that mean @abhinav and @bayesianmind are going to stamp? :1st_place_medal:

alsamylkin commented on app.go:
```diff
@@ -193,9 +198,43 @@ func New(opts ...Option) *App {
+func (app *App) executeInvokes() error {
+	for _, fn := range app.invokes {
+		fname := fxreflect.FuncName(fn)
+		app.logger.Printf("INVOKE\t\t%s", fname)
```
should we log the invoke after type check for the Option?

bayesianmind commented on app.go:
```diff
@@ -193,9 +198,55 @@ func New(opts ...Option) *App {
+		}
+
+		if err != nil {
+			app.logger.Printf("Error during %q invoke: %v", fname, err)
```
This is bugged, you need a break here. Otherwise the next invoke will just nuke the err.

abhinav replied:
+1

glibsm replied:
Will do.

glibsm replied:
Fixed. Added a test to cover, too

abhinav commented on app.go:
```diff
@@ -193,9 +198,55 @@ func New(opts ...Option) *App {
+//
+// This includes things like incomplete graphs, circular dependencies,
+// missing dependencies, invalid constructors, etc.
+func (app *App) Err() error {
```
Actually, this also includes errors encountered during invokes.

abhinav commented on app.go:
```diff
@@ -193,9 +198,55 @@ func New(opts ...Option) *App {
+// It might be worthwhile to consider adding context.Context to this function
+// so we can handle the infinite-invokes.
+//
+// Returns the first error encountered. Maybe shoul multiErr them together.
```
outdated doc?

glibsm commented:
@bayesianmind I kept the `gsm-new-invokes` branch around in case you used that commit hash. Let me know when you can swap back to master.
